### PR TITLE
Hugging Face Integration: GPT2

### PR DESCRIPTION
**Pull request by emilydinan**
**Patch description**
I'm working on a hugging face integration for generative transformers. Initial implementation is for GPT2, but with the goal to expand to other available models.

For example, running the command `python examples/interactive.py -m hugging_face/gpt2 --inference beam --beam-size 10 --beam-block-ngram 3 --beam-context-block-ngram 3` yields the following conversation:

```
Enter Your Message: hi
[Gpt2]: .

I don't know what's going on," he said. "I'm just trying to figure it out."
Enter Your Message: Ok cool. What is your name?
[Gpt2]:

He didn't know. He didn't even know who he was. He had no idea what he was talking about. He wasn't even sure where he was from. He couldn't even tell if he was a man or a woman. He just couldn't figure out what was going on. It was like he was trying to make sense of it all, but he couldn't get it out of his head. He was like, "Oh my God, I can't believe this is happening to me. I'm so sorry."

I'm sorry," I said. I didn't want to hurt him. I just wanted him to know that he was okay. I wanted to make sure he knew that he wasn't going to hurt anyone else. That he was going to be okay. That it was okay for him to be here. That there was nothing wrong with me. That I was a good person. That this was the right thing to do for him, and that he had the right to do what he wanted to do. But he didn't understand what I was saying. I wasn't trying to tell him that I was sorry. I was just telling him that it wasn't okay for me to hurt anybody else. It wasn't

### CODE REVIEW

File: parlai/agents/hugging_face/gpt2.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    def build_model(self, states=None):
+        """
+        Build and return model.
+        """
+        return HFGPT2Model(self.opt, self.dict)
+
+    def _v2t(self, vec):
+        """
+        Override because for GPT2, the start token = end token.
```
Comment (emilydinan) [inline parlai/agents/hugging_face/gpt2.py]: actually, we don't need to do this anymore, so i will get rid of this

File: parlai/agents/hugging_face/dict.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from parlai.core.dict import DictionaryAgent
+
+try:
+    from transformers import GPT2Tokenizer
```
Comment (stephenroller) [inline parlai/agents/hugging_face/dict.py]: is this right? I know they have their new `tokenizers` repo

Comment (emilydinan) [reply parlai/agents/hugging_face/dict.py]: hmm... i think so. unless the docs are outdated. this is what the generation example in their repo imports https://github.com/huggingface/transformers/blob/ea56d305be0000b3f54ea378df77f92867779ba5/examples/run_generation.py#L31

**Comment by stephenroller** [comment]
Feature request: being able to try different model sizes.

**Comment by emilydinan** [comment]
> Feature request: being able to try different model sizes.

done

File: parlai/agents/hugging_face/dict.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return {}
+
+
+class Gpt2DictionaryAgent(HuggingFaceDictionaryAgent):
+    def get_tokenizer(self, opt):
+        """
+        Instantiate tokenizer.
+        """
+        model_sz = opt['gpt2_size']
+        fle_key = 'gpt2' if model_sz == 'small' else f'gpt2-{model_sz}'
```
Comment (stephenroller) [inline parlai/agents/hugging_face/dict.py]: lol

File: parlai/agents/hugging_face/hugging_face.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+GPT2. To use this model, run with `-m hugging_face/gpt2`.
+"""
+try:
+    import transformers  # noqa: F401
+except ImportError as e:
+    raise e('Please run `pip install transformers`.')
+
+
+class HuggingFaceAgent:
+    raise RuntimeError(
```
Comment (stephenroller) [inline parlai/agents/hugging_face/hugging_face.py]: is this right?

**Comment by klshuster** [comment]
Just read the example output, golden

**Comment by emilydinan** [comment]
I still  have to add a README (and maybe a test?) but this is ready for review otherwise

File: parlai/agents/hugging_face/dict.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from parlai.core.dict import DictionaryAgent
+
+try:
+    from transformers import GPT2Tokenizer
+except ImportError as e:
+    raise e(
```
Comment (klshuster) [inline parlai/agents/hugging_face/dict.py]: is this actually broken? (i.e. the typing error)

File: parlai/agents/hugging_face/dict.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    def _define_special_tokens(self, opt):
+        if opt['add_special_tokens']:
+            # Add addtional start/end/pad tokens
+            self.tokenizer.add_special_tokens(SPECIAL_TOKENS)
+            self.start_token = SPECIAL_TOKENS["bos_token"]
+            self.end_token = SPECIAL_TOKENS["eos_token"]
+            self.null_token = SPECIAL_TOKENS["pad_token"]
+        else:
+            # Only special token is end of text
+            self.start_token = "x"  # hack, we cut off the start token
```
Comment (klshuster) [inline parlai/agents/hugging_face/dict.py]: nit: maybe a const. at the top? like `NO-OP` or something

File: parlai/agents/hugging_face/dict.py (hunk 5 above)
Comment (stephenroller) [inline parlai/agents/hugging_face/dict.py]: as mypy says: raise a new ImportError

File: parlai/agents/hugging_face/dict.py (hunk 2 above)
Comment (stephenroller) [reply parlai/agents/hugging_face/dict.py]: Okay, put it off for later

File: parlai/agents/hugging_face/dict.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    def get_tokenizer(self, opt):
+        """
+        Instantiate tokenizer.
+        """
+        model_sz = opt['gpt2_size']
+        fle_key = 'gpt2' if model_sz == 'small' else f'gpt2-{model_sz}'
+        return GPT2Tokenizer.from_pretrained(fle_key)
+
+    def _define_special_tokens(self, opt):
+        if opt['add_special_tokens']:
```
Comment (stephenroller) [inline parlai/agents/hugging_face/dict.py]: do we want a flag, or do we want to make an opinionated decision? We know it doesn't work well if you don't fine tune on dialogue and that's fine.

File: parlai/agents/hugging_face/gpt2.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+        self.config = self.decoder.transformer.config
+        self.lm_head = torch.nn.Linear(
+            self.config.n_embd, self.config.vocab_size, bias=False
+        )
+        self.tie_weights(self.lm_head, self.decoder.transformer.wte)
+
+        self.text_lengths = None  # used to reverse concatenation of context and labels
+
+    def tie_weights(self, output_embeddings, input_embeddings):
```
Comment (stephenroller) [inline parlai/agents/hugging_face/gpt2.py]: make this _private

File: parlai/agents/hugging_face/dict.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+from abc import ABC, abstractmethod
+
+SPECIAL_TOKENS = {
+    "bos_token": "<bos>",
+    "eos_token": "<eos>",
+    "pad_token": "<pad>",
+}
+
+
+class HuggingFaceDictionaryAgent(DictionaryAgent, ABC):
```
Comment (stephenroller) [inline parlai/agents/hugging_face/dict.py]: There's a general TODO for us to be aware of later in the half. DictionaryAgent should become a truly abstract, and then we should take the current DA and call it a ParlaiDictionaryAgent and have all our years of grossness into it.

Definitely not for now

File: parlai/agents/hugging_face/gpt2.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            total_length = max(self.text_lengths)
+            to_select = tensor.size(1) - total_length + 1
+            if to_select > 0:
+                # select only label scores
+                bsz = tensor.size(0)
+                new_tensors = []
+                for i in range(bsz):
+                    start = self.text_lengths[i] - 1
+                    end = start + to_select
+                    new_tensors.append(tensor[i, start:end, :].unsqueeze(0))
```
Comment (klshuster) [inline parlai/agents/hugging_face/gpt2.py]: protip: you can do `tensor[i:i+1, start:end, :]` and leave out the `unsqueeze(0)`

File: parlai/agents/hugging_face/gpt2.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+from transformers import GPT2Model
+
+import torch
+
+############################################
+## Modules
+############################################
+
+
+class GPT2Decoder(torch.nn.Module):
```
Comment (stephenroller) [inline parlai/agents/hugging_face/gpt2.py]: a very brief 2 sentences on the high level (We use a dummy encoder because ...) *could* be helpful at the top here.

File: parlai/agents/hugging_face/gpt2.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+            help='Add special tokens (like PAD, etc.). If False, '
+            'Can only use with batch size 1.',
+        )
+        argparser.set_defaults(
+            text_truncate=768,
+            label_truncate=256,
+            dict_maxexs=0,  # skip building dictionary
+        )
+        super(Gpt2Agent, cls).add_cmdline_args(argparser)
+        warn_once('WARNING: this model is in beta and the API is ' 'subject to change.')
```
Comment (klshuster) [inline parlai/agents/hugging_face/gpt2.py]: nit: you can remove the extra `'` here since it looks like black combined to one line

File: parlai/agents/hugging_face/gpt2.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        return agent
+
+    def __init__(self, opt, shared=None):
+        if not opt['add_special_tokens'] and opt['batchsize'] > 1:
+            raise RuntimeError(
+                'If using batchsize > 1, --add-special-tokens ' 'must be True.'
```
Comment (klshuster) [inline parlai/agents/hugging_face/gpt2.py]: ditto

File: parlai/agents/hugging_face/gpt2.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+                    start = self.text_lengths[i] - 1
+                    end = start + to_select
+                    new_tensors.append(tensor[i, start:end, :].unsqueeze(0))
+                tensor = torch.cat(new_tensors, 0)
+
+        return self.lm_head(tensor)
+
+    def reorder_decoder_incremental_state(self, incremental_state, inds):
+        new_incr_state = []
+        for layer_past in incremental_state:
```
Comment (stephenroller) [inline parlai/agents/hugging_face/gpt2.py]: I don't *know*, but this seems more expensive then it needs to be? why not a `torch.index_select` on each layer? 

The clone/detach is also particularly worrisome -- this would cause subtle issues if we did RL here and tried backpropping based on some RL signal.

File: parlai/agents/hugging_face/hugging_face.py (hunk 4 above)
Comment (stephenroller) [reply parlai/agents/hugging_face/hugging_face.py]: it's definitely not though.  It needs to be in `__init__`

File: parlai/utils/torch.py [hunk 15]
```diff
@@ -235,3 +268,15 @@ def fp16_available() -> bool:
             'install APEX from https://github.com/NVIDIA/apex.'
         )
         return False
+
+
+class DummyEncoder(torch.nn.Module):
```
Comment (stephenroller) [inline parlai/utils/torch.py]: If we're really going to put this into a util folder, I'd prefer it be called IdentityLayer.

**Comment by stephenroller** [comment]
to fix CI, add `pip install transformers` into the gpu setup and increase the version number for the cache.

**Comment by stephenroller** [comment]
Also don't forget the beta warning we talked about

File: parlai/agents/hugging_face/gpt2.py (hunk 14 above)
Comment (emilydinan) [reply parlai/agents/hugging_face/gpt2.py]: I just copied what they did, but can chnage

**Comment by emilydinan** [comment]
> Also don't forget the beta warning we talked about

It is already there?

File: parlai/agents/hugging_face/dict.py (hunk 7 above)
Comment (emilydinan) [reply parlai/agents/hugging_face/dict.py]: No, I want a flag. For finetuning, we need the special tokens. But in order to evaluate teh model straight out of the box, we cannot add them (other wise PPL --> infinity)

File: parlai/agents/hugging_face/dict.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        pass
+
+    def txt2vec(self, text, vec_type=list):
+        tokens = self.tokenizer.tokenize(text)
+        tokens_id = self.tokenizer.convert_tokens_to_ids(tokens)
+        return tokens_id
+
+    def vec2txt(self, vec):
+        return self.tokenizer.decode(vec, clean_up_tokenization_spaces=True)
```
[the remark this answers is not in the document]
Comment (emilydinan) [reply parlai/agents/hugging_face/dict.py]: according to the HF docs:
"Clean up a list of simple English tokenization artifacts like spaces before punctuations and abreviated forms."
